### PR TITLE
Return StatusOr<void> from Client::DeleteObject.

### LINKED ISSUE #1664: Implement noex version of storage::Client::DeleteObject
The signature should be something like:

```C++
class storage::noex::Client { public:
  template <typename... Options>
  StatusOr<void> DeleteObject(std::string const& bucket_name,
                    std::string const& object_name, Options&&... options;
};
```

We need the code and a basic integration test. The existing tests should cover more complicated cases, but use your judgement.

### PR DESCRIPTION
**Pull request by coryan**
Changed tests and examples to check the return status.

This fixes #1664.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1841)

### CODE REVIEW

File: ci/test-install/storage_install_test.cc [hunk 1]
```diff
@@ -47,8 +47,10 @@ int main(int argc, char* argv[]) try {
   while (std::getline(stream, line, '\n')) {
     ++count;
   }
-  client.DeleteObject(bucket_name, object_name,
-                      gcs::Generation(meta.generation()));
+  client
+      .DeleteObject(bucket_name, object_name,
+                    gcs::Generation(meta.generation()))
+      .value();
```
Comment (frankyn) [inline ci/test-install/storage_install_test.cc]: This is a little confusing as the value is not used. Is this meant to throw if the DeleteObject request isn't successful?

Comment (coryan) [reply ci/test-install/storage_install_test.cc]: `DeleteObject` returns a `StatusOr<void>`, `.value()` does throw if there is an error, but there is no way to "use" the `void` value.  The rest of the code uses exceptions to signal errors, If we move to use explicit error checking that should be in a separate PR.

**Comment by coryan** [comment]
PTAL.

File: ci/test-install/storage_install_test.cc (hunk 1 above)
Comment (frankyn) [reply ci/test-install/storage_install_test.cc]: Sg, thanks for clarifying.

**Pull request merged**